### PR TITLE
abstract_tcp_server2: guard against negative timeouts

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

File: contrib/epee/include/net/abstract_tcp_server2.inl [hunk 1]
```diff
@@ -721,7 +721,9 @@ PRAGMA_WARNING_DISABLE_VS(4355)
   boost::posix_time::milliseconds connection<t_protocol_handler>::get_timeout_from_bytes_read(size_t bytes)
   {
     boost::posix_time::milliseconds ms = (boost::posix_time::milliseconds)(unsigned)(bytes * TIMEOUT_EXTRA_MS_PER_BYTE);
-    ms += m_timer.expires_from_now();
+    const auto cur = m_timer.expires_from_now().total_milliseconds();
+    if (cur > 0)
+      ms += (boost::posix_time::milliseconds)cur;
```
Comment (vtnerd) [inline contrib/epee/include/net/abstract_tcp_server2.inl]: This is a:

```c++
ms += std::max(m_timer.expires_from_now(), boost::posix_time::milliseconds(0));
```

File: contrib/epee/include/net/abstract_tcp_server2.inl [hunk 2]
```diff
@@ -760,7 +767,11 @@ PRAGMA_WARNING_DISABLE_VS(4355)
       return;
     }
     if (add)
-      ms += m_timer.expires_from_now();
+    {
+      const auto cur = m_timer.expires_from_now().total_milliseconds();
```
Comment (vtnerd) [inline contrib/epee/include/net/abstract_tcp_server2.inl]: Same here as above.

File: contrib/epee/include/net/abstract_tcp_server2.inl [hunk 3]
```diff
@@ -363,8 +363,8 @@ PRAGMA_WARNING_DISABLE_VS(4355)
 				}
 				
 				delay *= 0.5;
-				if (delay > 0) {
-					long int ms = (long int)(delay * 100);
+				long int ms = (long int)(delay * 100);
```
Comment (vtnerd) [inline contrib/epee/include/net/abstract_tcp_server2.inl]: How was this going negative previously? Overflow?

Comment (moneromooo-monero) [reply contrib/epee/include/net/abstract_tcp_server2.inl]: I don't know for sure. binaryFate reported a negative delay, but I'm not sure exactly when it was from, so I armoured the places that call the delay. I made a program to check the ranges with those types and this seems it should be OK, if the throttle code returns a sane number, but I decided to armour anyway.

**Pull request merged**